


<!DOCTYPE html>
<html>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    
    
    <meta http-equiv="X-UA-Compatible" content="IE=emulateIE7" />
    <title>Coverage for /home/ubuntu/scratch/biodemo_dev/lib/python3.6/site-packages/Bio/AlignIO/__init__.py: 19%</title>
    <link rel="stylesheet" href="style.css" type="text/css">
    
    <script type="text/javascript" src="jquery.min.js"></script>
    <script type="text/javascript" src="jquery.hotkeys.js"></script>
    <script type="text/javascript" src="jquery.isonscreen.js"></script>
    <script type="text/javascript" src="coverage_html.js"></script>
    <script type="text/javascript">
        jQuery(document).ready(coverage.pyfile_ready);
    </script>
</head>
<body class="pyfile">

<div id="header">
    <div class="content">
        <h1>Coverage for <b>/home/ubuntu/scratch/biodemo_dev/lib/python3.6/site-packages/Bio/AlignIO/__init__.py</b> :
            <span class="pc_cov">19%</span>
        </h1>

        <img id="keyboard_icon" src="keybd_closed.png" alt="Show keyboard shortcuts" />

        <h2 class="stats">
            124 statements &nbsp;
            <span class="run hide_run shortkey_r button_toggle_run">23 run</span>
            <span class="mis shortkey_m button_toggle_mis">101 missing</span>
            <span class="exc shortkey_x button_toggle_exc">0 excluded</span>

            
        </h2>
    </div>
</div>

<div class="help_panel">
    <img id="panel_icon" src="keybd_open.png" alt="Hide keyboard shortcuts" />
    <p class="legend">Hot-keys on this page</p>
    <div>
    <p class="keyhelp">
        <span class="key">r</span>
        <span class="key">m</span>
        <span class="key">x</span>
        <span class="key">p</span> &nbsp; toggle line displays
    </p>
    <p class="keyhelp">
        <span class="key">j</span>
        <span class="key">k</span> &nbsp; next/prev highlighted chunk
    </p>
    <p class="keyhelp">
        <span class="key">0</span> &nbsp; (zero) top of page
    </p>
    <p class="keyhelp">
        <span class="key">1</span> &nbsp; (one) first highlighted chunk
    </p>
    </div>
</div>

<div id="source">
    <table>
        <tr>
            <td class="linenos">
<p id="n1" class="pln"><a href="#n1">1</a></p>
<p id="n2" class="pln"><a href="#n2">2</a></p>
<p id="n3" class="pln"><a href="#n3">3</a></p>
<p id="n4" class="pln"><a href="#n4">4</a></p>
<p id="n5" class="pln"><a href="#n5">5</a></p>
<p id="n6" class="pln"><a href="#n6">6</a></p>
<p id="n7" class="pln"><a href="#n7">7</a></p>
<p id="n8" class="stm run hide_run"><a href="#n8">8</a></p>
<p id="n9" class="pln"><a href="#n9">9</a></p>
<p id="n10" class="pln"><a href="#n10">10</a></p>
<p id="n11" class="pln"><a href="#n11">11</a></p>
<p id="n12" class="pln"><a href="#n12">12</a></p>
<p id="n13" class="pln"><a href="#n13">13</a></p>
<p id="n14" class="pln"><a href="#n14">14</a></p>
<p id="n15" class="pln"><a href="#n15">15</a></p>
<p id="n16" class="pln"><a href="#n16">16</a></p>
<p id="n17" class="pln"><a href="#n17">17</a></p>
<p id="n18" class="pln"><a href="#n18">18</a></p>
<p id="n19" class="pln"><a href="#n19">19</a></p>
<p id="n20" class="pln"><a href="#n20">20</a></p>
<p id="n21" class="pln"><a href="#n21">21</a></p>
<p id="n22" class="pln"><a href="#n22">22</a></p>
<p id="n23" class="pln"><a href="#n23">23</a></p>
<p id="n24" class="pln"><a href="#n24">24</a></p>
<p id="n25" class="pln"><a href="#n25">25</a></p>
<p id="n26" class="pln"><a href="#n26">26</a></p>
<p id="n27" class="pln"><a href="#n27">27</a></p>
<p id="n28" class="pln"><a href="#n28">28</a></p>
<p id="n29" class="pln"><a href="#n29">29</a></p>
<p id="n30" class="pln"><a href="#n30">30</a></p>
<p id="n31" class="pln"><a href="#n31">31</a></p>
<p id="n32" class="pln"><a href="#n32">32</a></p>
<p id="n33" class="pln"><a href="#n33">33</a></p>
<p id="n34" class="pln"><a href="#n34">34</a></p>
<p id="n35" class="pln"><a href="#n35">35</a></p>
<p id="n36" class="pln"><a href="#n36">36</a></p>
<p id="n37" class="pln"><a href="#n37">37</a></p>
<p id="n38" class="pln"><a href="#n38">38</a></p>
<p id="n39" class="pln"><a href="#n39">39</a></p>
<p id="n40" class="pln"><a href="#n40">40</a></p>
<p id="n41" class="pln"><a href="#n41">41</a></p>
<p id="n42" class="pln"><a href="#n42">42</a></p>
<p id="n43" class="pln"><a href="#n43">43</a></p>
<p id="n44" class="pln"><a href="#n44">44</a></p>
<p id="n45" class="pln"><a href="#n45">45</a></p>
<p id="n46" class="pln"><a href="#n46">46</a></p>
<p id="n47" class="pln"><a href="#n47">47</a></p>
<p id="n48" class="pln"><a href="#n48">48</a></p>
<p id="n49" class="pln"><a href="#n49">49</a></p>
<p id="n50" class="pln"><a href="#n50">50</a></p>
<p id="n51" class="pln"><a href="#n51">51</a></p>
<p id="n52" class="pln"><a href="#n52">52</a></p>
<p id="n53" class="pln"><a href="#n53">53</a></p>
<p id="n54" class="pln"><a href="#n54">54</a></p>
<p id="n55" class="pln"><a href="#n55">55</a></p>
<p id="n56" class="pln"><a href="#n56">56</a></p>
<p id="n57" class="pln"><a href="#n57">57</a></p>
<p id="n58" class="pln"><a href="#n58">58</a></p>
<p id="n59" class="pln"><a href="#n59">59</a></p>
<p id="n60" class="pln"><a href="#n60">60</a></p>
<p id="n61" class="pln"><a href="#n61">61</a></p>
<p id="n62" class="pln"><a href="#n62">62</a></p>
<p id="n63" class="pln"><a href="#n63">63</a></p>
<p id="n64" class="pln"><a href="#n64">64</a></p>
<p id="n65" class="pln"><a href="#n65">65</a></p>
<p id="n66" class="pln"><a href="#n66">66</a></p>
<p id="n67" class="pln"><a href="#n67">67</a></p>
<p id="n68" class="pln"><a href="#n68">68</a></p>
<p id="n69" class="pln"><a href="#n69">69</a></p>
<p id="n70" class="pln"><a href="#n70">70</a></p>
<p id="n71" class="pln"><a href="#n71">71</a></p>
<p id="n72" class="pln"><a href="#n72">72</a></p>
<p id="n73" class="pln"><a href="#n73">73</a></p>
<p id="n74" class="pln"><a href="#n74">74</a></p>
<p id="n75" class="pln"><a href="#n75">75</a></p>
<p id="n76" class="pln"><a href="#n76">76</a></p>
<p id="n77" class="pln"><a href="#n77">77</a></p>
<p id="n78" class="pln"><a href="#n78">78</a></p>
<p id="n79" class="pln"><a href="#n79">79</a></p>
<p id="n80" class="pln"><a href="#n80">80</a></p>
<p id="n81" class="pln"><a href="#n81">81</a></p>
<p id="n82" class="pln"><a href="#n82">82</a></p>
<p id="n83" class="pln"><a href="#n83">83</a></p>
<p id="n84" class="pln"><a href="#n84">84</a></p>
<p id="n85" class="pln"><a href="#n85">85</a></p>
<p id="n86" class="pln"><a href="#n86">86</a></p>
<p id="n87" class="pln"><a href="#n87">87</a></p>
<p id="n88" class="pln"><a href="#n88">88</a></p>
<p id="n89" class="pln"><a href="#n89">89</a></p>
<p id="n90" class="pln"><a href="#n90">90</a></p>
<p id="n91" class="pln"><a href="#n91">91</a></p>
<p id="n92" class="pln"><a href="#n92">92</a></p>
<p id="n93" class="pln"><a href="#n93">93</a></p>
<p id="n94" class="pln"><a href="#n94">94</a></p>
<p id="n95" class="pln"><a href="#n95">95</a></p>
<p id="n96" class="pln"><a href="#n96">96</a></p>
<p id="n97" class="pln"><a href="#n97">97</a></p>
<p id="n98" class="pln"><a href="#n98">98</a></p>
<p id="n99" class="pln"><a href="#n99">99</a></p>
<p id="n100" class="pln"><a href="#n100">100</a></p>
<p id="n101" class="pln"><a href="#n101">101</a></p>
<p id="n102" class="pln"><a href="#n102">102</a></p>
<p id="n103" class="pln"><a href="#n103">103</a></p>
<p id="n104" class="pln"><a href="#n104">104</a></p>
<p id="n105" class="pln"><a href="#n105">105</a></p>
<p id="n106" class="pln"><a href="#n106">106</a></p>
<p id="n107" class="pln"><a href="#n107">107</a></p>
<p id="n108" class="pln"><a href="#n108">108</a></p>
<p id="n109" class="pln"><a href="#n109">109</a></p>
<p id="n110" class="pln"><a href="#n110">110</a></p>
<p id="n111" class="pln"><a href="#n111">111</a></p>
<p id="n112" class="pln"><a href="#n112">112</a></p>
<p id="n113" class="pln"><a href="#n113">113</a></p>
<p id="n114" class="pln"><a href="#n114">114</a></p>
<p id="n115" class="pln"><a href="#n115">115</a></p>
<p id="n116" class="pln"><a href="#n116">116</a></p>
<p id="n117" class="pln"><a href="#n117">117</a></p>
<p id="n118" class="pln"><a href="#n118">118</a></p>
<p id="n119" class="pln"><a href="#n119">119</a></p>
<p id="n120" class="pln"><a href="#n120">120</a></p>
<p id="n121" class="pln"><a href="#n121">121</a></p>
<p id="n122" class="pln"><a href="#n122">122</a></p>
<p id="n123" class="pln"><a href="#n123">123</a></p>
<p id="n124" class="pln"><a href="#n124">124</a></p>
<p id="n125" class="pln"><a href="#n125">125</a></p>
<p id="n126" class="pln"><a href="#n126">126</a></p>
<p id="n127" class="pln"><a href="#n127">127</a></p>
<p id="n128" class="pln"><a href="#n128">128</a></p>
<p id="n129" class="pln"><a href="#n129">129</a></p>
<p id="n130" class="stm run hide_run"><a href="#n130">130</a></p>
<p id="n131" class="stm run hide_run"><a href="#n131">131</a></p>
<p id="n132" class="pln"><a href="#n132">132</a></p>
<p id="n133" class="pln"><a href="#n133">133</a></p>
<p id="n134" class="pln"><a href="#n134">134</a></p>
<p id="n135" class="pln"><a href="#n135">135</a></p>
<p id="n136" class="pln"><a href="#n136">136</a></p>
<p id="n137" class="pln"><a href="#n137">137</a></p>
<p id="n138" class="pln"><a href="#n138">138</a></p>
<p id="n139" class="pln"><a href="#n139">139</a></p>
<p id="n140" class="pln"><a href="#n140">140</a></p>
<p id="n141" class="pln"><a href="#n141">141</a></p>
<p id="n142" class="pln"><a href="#n142">142</a></p>
<p id="n143" class="pln"><a href="#n143">143</a></p>
<p id="n144" class="pln"><a href="#n144">144</a></p>
<p id="n145" class="pln"><a href="#n145">145</a></p>
<p id="n146" class="pln"><a href="#n146">146</a></p>
<p id="n147" class="stm run hide_run"><a href="#n147">147</a></p>
<p id="n148" class="stm run hide_run"><a href="#n148">148</a></p>
<p id="n149" class="stm run hide_run"><a href="#n149">149</a></p>
<p id="n150" class="pln"><a href="#n150">150</a></p>
<p id="n151" class="stm run hide_run"><a href="#n151">151</a></p>
<p id="n152" class="stm run hide_run"><a href="#n152">152</a></p>
<p id="n153" class="stm run hide_run"><a href="#n153">153</a></p>
<p id="n154" class="stm run hide_run"><a href="#n154">154</a></p>
<p id="n155" class="stm run hide_run"><a href="#n155">155</a></p>
<p id="n156" class="stm run hide_run"><a href="#n156">156</a></p>
<p id="n157" class="stm run hide_run"><a href="#n157">157</a></p>
<p id="n158" class="stm run hide_run"><a href="#n158">158</a></p>
<p id="n159" class="pln"><a href="#n159">159</a></p>
<p id="n160" class="pln"><a href="#n160">160</a></p>
<p id="n161" class="pln"><a href="#n161">161</a></p>
<p id="n162" class="pln"><a href="#n162">162</a></p>
<p id="n163" class="stm run hide_run"><a href="#n163">163</a></p>
<p id="n164" class="pln"><a href="#n164">164</a></p>
<p id="n165" class="pln"><a href="#n165">165</a></p>
<p id="n166" class="pln"><a href="#n166">166</a></p>
<p id="n167" class="pln"><a href="#n167">167</a></p>
<p id="n168" class="pln"><a href="#n168">168</a></p>
<p id="n169" class="pln"><a href="#n169">169</a></p>
<p id="n170" class="pln"><a href="#n170">170</a></p>
<p id="n171" class="pln"><a href="#n171">171</a></p>
<p id="n172" class="pln"><a href="#n172">172</a></p>
<p id="n173" class="pln"><a href="#n173">173</a></p>
<p id="n174" class="pln"><a href="#n174">174</a></p>
<p id="n175" class="pln"><a href="#n175">175</a></p>
<p id="n176" class="stm run hide_run"><a href="#n176">176</a></p>
<p id="n177" class="pln"><a href="#n177">177</a></p>
<p id="n178" class="pln"><a href="#n178">178</a></p>
<p id="n179" class="pln"><a href="#n179">179</a></p>
<p id="n180" class="pln"><a href="#n180">180</a></p>
<p id="n181" class="pln"><a href="#n181">181</a></p>
<p id="n182" class="pln"><a href="#n182">182</a></p>
<p id="n183" class="pln"><a href="#n183">183</a></p>
<p id="n184" class="pln"><a href="#n184">184</a></p>
<p id="n185" class="pln"><a href="#n185">185</a></p>
<p id="n186" class="pln"><a href="#n186">186</a></p>
<p id="n187" class="pln"><a href="#n187">187</a></p>
<p id="n188" class="pln"><a href="#n188">188</a></p>
<p id="n189" class="stm run hide_run"><a href="#n189">189</a></p>
<p id="n190" class="pln"><a href="#n190">190</a></p>
<p id="n191" class="pln"><a href="#n191">191</a></p>
<p id="n192" class="pln"><a href="#n192">192</a></p>
<p id="n193" class="pln"><a href="#n193">193</a></p>
<p id="n194" class="pln"><a href="#n194">194</a></p>
<p id="n195" class="pln"><a href="#n195">195</a></p>
<p id="n196" class="pln"><a href="#n196">196</a></p>
<p id="n197" class="pln"><a href="#n197">197</a></p>
<p id="n198" class="pln"><a href="#n198">198</a></p>
<p id="n199" class="pln"><a href="#n199">199</a></p>
<p id="n200" class="pln"><a href="#n200">200</a></p>
<p id="n201" class="pln"><a href="#n201">201</a></p>
<p id="n202" class="pln"><a href="#n202">202</a></p>
<p id="n203" class="stm mis"><a href="#n203">203</a></p>
<p id="n204" class="pln"><a href="#n204">204</a></p>
<p id="n205" class="pln"><a href="#n205">205</a></p>
<p id="n206" class="stm mis"><a href="#n206">206</a></p>
<p id="n207" class="stm mis"><a href="#n207">207</a></p>
<p id="n208" class="stm mis"><a href="#n208">208</a></p>
<p id="n209" class="stm mis"><a href="#n209">209</a></p>
<p id="n210" class="stm mis"><a href="#n210">210</a></p>
<p id="n211" class="stm mis"><a href="#n211">211</a></p>
<p id="n212" class="pln"><a href="#n212">212</a></p>
<p id="n213" class="stm mis"><a href="#n213">213</a></p>
<p id="n214" class="pln"><a href="#n214">214</a></p>
<p id="n215" class="stm mis"><a href="#n215">215</a></p>
<p id="n216" class="pln"><a href="#n216">216</a></p>
<p id="n217" class="stm mis"><a href="#n217">217</a></p>
<p id="n218" class="pln"><a href="#n218">218</a></p>
<p id="n219" class="stm mis"><a href="#n219">219</a></p>
<p id="n220" class="stm mis"><a href="#n220">220</a></p>
<p id="n221" class="stm mis"><a href="#n221">221</a></p>
<p id="n222" class="stm mis"><a href="#n222">222</a></p>
<p id="n223" class="pln"><a href="#n223">223</a></p>
<p id="n224" class="pln"><a href="#n224">224</a></p>
<p id="n225" class="stm mis"><a href="#n225">225</a></p>
<p id="n226" class="stm mis"><a href="#n226">226</a></p>
<p id="n227" class="stm mis"><a href="#n227">227</a></p>
<p id="n228" class="stm mis"><a href="#n228">228</a></p>
<p id="n229" class="pln"><a href="#n229">229</a></p>
<p id="n230" class="stm mis"><a href="#n230">230</a></p>
<p id="n231" class="stm mis"><a href="#n231">231</a></p>
<p id="n232" class="stm mis"><a href="#n232">232</a></p>
<p id="n233" class="stm mis"><a href="#n233">233</a></p>
<p id="n234" class="pln"><a href="#n234">234</a></p>
<p id="n235" class="pln"><a href="#n235">235</a></p>
<p id="n236" class="stm mis"><a href="#n236">236</a></p>
<p id="n237" class="pln"><a href="#n237">237</a></p>
<p id="n238" class="stm mis"><a href="#n238">238</a></p>
<p id="n239" class="stm mis"><a href="#n239">239</a></p>
<p id="n240" class="pln"><a href="#n240">240</a></p>
<p id="n241" class="pln"><a href="#n241">241</a></p>
<p id="n242" class="pln"><a href="#n242">242</a></p>
<p id="n243" class="stm mis"><a href="#n243">243</a></p>
<p id="n244" class="pln"><a href="#n244">244</a></p>
<p id="n245" class="pln"><a href="#n245">245</a></p>
<p id="n246" class="pln"><a href="#n246">246</a></p>
<p id="n247" class="stm run hide_run"><a href="#n247">247</a></p>
<p id="n248" class="pln"><a href="#n248">248</a></p>
<p id="n249" class="pln"><a href="#n249">249</a></p>
<p id="n250" class="pln"><a href="#n250">250</a></p>
<p id="n251" class="pln"><a href="#n251">251</a></p>
<p id="n252" class="pln"><a href="#n252">252</a></p>
<p id="n253" class="pln"><a href="#n253">253</a></p>
<p id="n254" class="pln"><a href="#n254">254</a></p>
<p id="n255" class="pln"><a href="#n255">255</a></p>
<p id="n256" class="pln"><a href="#n256">256</a></p>
<p id="n257" class="pln"><a href="#n257">257</a></p>
<p id="n258" class="pln"><a href="#n258">258</a></p>
<p id="n259" class="pln"><a href="#n259">259</a></p>
<p id="n260" class="pln"><a href="#n260">260</a></p>
<p id="n261" class="pln"><a href="#n261">261</a></p>
<p id="n262" class="stm mis"><a href="#n262">262</a></p>
<p id="n263" class="pln"><a href="#n263">263</a></p>
<p id="n264" class="stm mis"><a href="#n264">264</a></p>
<p id="n265" class="stm mis"><a href="#n265">265</a></p>
<p id="n266" class="pln"><a href="#n266">266</a></p>
<p id="n267" class="stm mis"><a href="#n267">267</a></p>
<p id="n268" class="pln"><a href="#n268">268</a></p>
<p id="n269" class="stm mis"><a href="#n269">269</a></p>
<p id="n270" class="pln"><a href="#n270">270</a></p>
<p id="n271" class="stm mis"><a href="#n271">271</a></p>
<p id="n272" class="stm mis"><a href="#n272">272</a></p>
<p id="n273" class="stm mis"><a href="#n273">273</a></p>
<p id="n274" class="stm mis"><a href="#n274">274</a></p>
<p id="n275" class="stm mis"><a href="#n275">275</a></p>
<p id="n276" class="stm mis"><a href="#n276">276</a></p>
<p id="n277" class="stm mis"><a href="#n277">277</a></p>
<p id="n278" class="stm mis"><a href="#n278">278</a></p>
<p id="n279" class="pln"><a href="#n279">279</a></p>
<p id="n280" class="pln"><a href="#n280">280</a></p>
<p id="n281" class="pln"><a href="#n281">281</a></p>
<p id="n282" class="stm mis"><a href="#n282">282</a></p>
<p id="n283" class="stm mis"><a href="#n283">283</a></p>
<p id="n284" class="stm mis"><a href="#n284">284</a></p>
<p id="n285" class="pln"><a href="#n285">285</a></p>
<p id="n286" class="pln"><a href="#n286">286</a></p>
<p id="n287" class="stm run hide_run"><a href="#n287">287</a></p>
<p id="n288" class="pln"><a href="#n288">288</a></p>
<p id="n289" class="pln"><a href="#n289">289</a></p>
<p id="n290" class="stm mis"><a href="#n290">290</a></p>
<p id="n291" class="stm mis"><a href="#n291">291</a></p>
<p id="n292" class="stm mis"><a href="#n292">292</a></p>
<p id="n293" class="pln"><a href="#n293">293</a></p>
<p id="n294" class="stm mis"><a href="#n294">294</a></p>
<p id="n295" class="pln"><a href="#n295">295</a></p>
<p id="n296" class="pln"><a href="#n296">296</a></p>
<p id="n297" class="stm mis"><a href="#n297">297</a></p>
<p id="n298" class="stm mis"><a href="#n298">298</a></p>
<p id="n299" class="pln"><a href="#n299">299</a></p>
<p id="n300" class="stm mis"><a href="#n300">300</a></p>
<p id="n301" class="pln"><a href="#n301">301</a></p>
<p id="n302" class="pln"><a href="#n302">302</a></p>
<p id="n303" class="stm mis"><a href="#n303">303</a></p>
<p id="n304" class="stm mis"><a href="#n304">304</a></p>
<p id="n305" class="stm mis"><a href="#n305">305</a></p>
<p id="n306" class="pln"><a href="#n306">306</a></p>
<p id="n307" class="pln"><a href="#n307">307</a></p>
<p id="n308" class="stm run hide_run"><a href="#n308">308</a></p>
<p id="n309" class="pln"><a href="#n309">309</a></p>
<p id="n310" class="pln"><a href="#n310">310</a></p>
<p id="n311" class="pln"><a href="#n311">311</a></p>
<p id="n312" class="pln"><a href="#n312">312</a></p>
<p id="n313" class="pln"><a href="#n313">313</a></p>
<p id="n314" class="pln"><a href="#n314">314</a></p>
<p id="n315" class="pln"><a href="#n315">315</a></p>
<p id="n316" class="pln"><a href="#n316">316</a></p>
<p id="n317" class="pln"><a href="#n317">317</a></p>
<p id="n318" class="pln"><a href="#n318">318</a></p>
<p id="n319" class="pln"><a href="#n319">319</a></p>
<p id="n320" class="pln"><a href="#n320">320</a></p>
<p id="n321" class="pln"><a href="#n321">321</a></p>
<p id="n322" class="pln"><a href="#n322">322</a></p>
<p id="n323" class="pln"><a href="#n323">323</a></p>
<p id="n324" class="pln"><a href="#n324">324</a></p>
<p id="n325" class="pln"><a href="#n325">325</a></p>
<p id="n326" class="pln"><a href="#n326">326</a></p>
<p id="n327" class="pln"><a href="#n327">327</a></p>
<p id="n328" class="pln"><a href="#n328">328</a></p>
<p id="n329" class="pln"><a href="#n329">329</a></p>
<p id="n330" class="pln"><a href="#n330">330</a></p>
<p id="n331" class="pln"><a href="#n331">331</a></p>
<p id="n332" class="pln"><a href="#n332">332</a></p>
<p id="n333" class="pln"><a href="#n333">333</a></p>
<p id="n334" class="pln"><a href="#n334">334</a></p>
<p id="n335" class="pln"><a href="#n335">335</a></p>
<p id="n336" class="pln"><a href="#n336">336</a></p>
<p id="n337" class="pln"><a href="#n337">337</a></p>
<p id="n338" class="pln"><a href="#n338">338</a></p>
<p id="n339" class="pln"><a href="#n339">339</a></p>
<p id="n340" class="pln"><a href="#n340">340</a></p>
<p id="n341" class="pln"><a href="#n341">341</a></p>
<p id="n342" class="stm mis"><a href="#n342">342</a></p>
<p id="n343" class="pln"><a href="#n343">343</a></p>
<p id="n344" class="pln"><a href="#n344">344</a></p>
<p id="n345" class="stm mis"><a href="#n345">345</a></p>
<p id="n346" class="stm mis"><a href="#n346">346</a></p>
<p id="n347" class="stm mis"><a href="#n347">347</a></p>
<p id="n348" class="stm mis"><a href="#n348">348</a></p>
<p id="n349" class="stm mis"><a href="#n349">349</a></p>
<p id="n350" class="stm mis"><a href="#n350">350</a></p>
<p id="n351" class="stm mis"><a href="#n351">351</a></p>
<p id="n352" class="pln"><a href="#n352">352</a></p>
<p id="n353" class="stm mis"><a href="#n353">353</a></p>
<p id="n354" class="stm mis"><a href="#n354">354</a></p>
<p id="n355" class="stm mis"><a href="#n355">355</a></p>
<p id="n356" class="pln"><a href="#n356">356</a></p>
<p id="n357" class="stm mis"><a href="#n357">357</a></p>
<p id="n358" class="pln"><a href="#n358">358</a></p>
<p id="n359" class="stm mis"><a href="#n359">359</a></p>
<p id="n360" class="stm mis"><a href="#n360">360</a></p>
<p id="n361" class="stm mis"><a href="#n361">361</a></p>
<p id="n362" class="stm mis"><a href="#n362">362</a></p>
<p id="n363" class="pln"><a href="#n363">363</a></p>
<p id="n364" class="stm mis"><a href="#n364">364</a></p>
<p id="n365" class="pln"><a href="#n365">365</a></p>
<p id="n366" class="stm mis"><a href="#n366">366</a></p>
<p id="n367" class="stm mis"><a href="#n367">367</a></p>
<p id="n368" class="pln"><a href="#n368">368</a></p>
<p id="n369" class="stm mis"><a href="#n369">369</a></p>
<p id="n370" class="pln"><a href="#n370">370</a></p>
<p id="n371" class="pln"><a href="#n371">371</a></p>
<p id="n372" class="stm mis"><a href="#n372">372</a></p>
<p id="n373" class="pln"><a href="#n373">373</a></p>
<p id="n374" class="stm mis"><a href="#n374">374</a></p>
<p id="n375" class="pln"><a href="#n375">375</a></p>
<p id="n376" class="pln"><a href="#n376">376</a></p>
<p id="n377" class="pln"><a href="#n377">377</a></p>
<p id="n378" class="stm mis"><a href="#n378">378</a></p>
<p id="n379" class="pln"><a href="#n379">379</a></p>
<p id="n380" class="pln"><a href="#n380">380</a></p>
<p id="n381" class="pln"><a href="#n381">381</a></p>
<p id="n382" class="stm mis"><a href="#n382">382</a></p>
<p id="n383" class="stm mis"><a href="#n383">383</a></p>
<p id="n384" class="pln"><a href="#n384">384</a></p>
<p id="n385" class="pln"><a href="#n385">385</a></p>
<p id="n386" class="stm run hide_run"><a href="#n386">386</a></p>
<p id="n387" class="pln"><a href="#n387">387</a></p>
<p id="n388" class="pln"><a href="#n388">388</a></p>
<p id="n389" class="pln"><a href="#n389">389</a></p>
<p id="n390" class="pln"><a href="#n390">390</a></p>
<p id="n391" class="pln"><a href="#n391">391</a></p>
<p id="n392" class="pln"><a href="#n392">392</a></p>
<p id="n393" class="pln"><a href="#n393">393</a></p>
<p id="n394" class="pln"><a href="#n394">394</a></p>
<p id="n395" class="pln"><a href="#n395">395</a></p>
<p id="n396" class="pln"><a href="#n396">396</a></p>
<p id="n397" class="pln"><a href="#n397">397</a></p>
<p id="n398" class="pln"><a href="#n398">398</a></p>
<p id="n399" class="pln"><a href="#n399">399</a></p>
<p id="n400" class="pln"><a href="#n400">400</a></p>
<p id="n401" class="pln"><a href="#n401">401</a></p>
<p id="n402" class="pln"><a href="#n402">402</a></p>
<p id="n403" class="pln"><a href="#n403">403</a></p>
<p id="n404" class="pln"><a href="#n404">404</a></p>
<p id="n405" class="pln"><a href="#n405">405</a></p>
<p id="n406" class="pln"><a href="#n406">406</a></p>
<p id="n407" class="pln"><a href="#n407">407</a></p>
<p id="n408" class="pln"><a href="#n408">408</a></p>
<p id="n409" class="pln"><a href="#n409">409</a></p>
<p id="n410" class="pln"><a href="#n410">410</a></p>
<p id="n411" class="pln"><a href="#n411">411</a></p>
<p id="n412" class="pln"><a href="#n412">412</a></p>
<p id="n413" class="pln"><a href="#n413">413</a></p>
<p id="n414" class="pln"><a href="#n414">414</a></p>
<p id="n415" class="pln"><a href="#n415">415</a></p>
<p id="n416" class="pln"><a href="#n416">416</a></p>
<p id="n417" class="pln"><a href="#n417">417</a></p>
<p id="n418" class="pln"><a href="#n418">418</a></p>
<p id="n419" class="pln"><a href="#n419">419</a></p>
<p id="n420" class="pln"><a href="#n420">420</a></p>
<p id="n421" class="pln"><a href="#n421">421</a></p>
<p id="n422" class="pln"><a href="#n422">422</a></p>
<p id="n423" class="pln"><a href="#n423">423</a></p>
<p id="n424" class="pln"><a href="#n424">424</a></p>
<p id="n425" class="pln"><a href="#n425">425</a></p>
<p id="n426" class="pln"><a href="#n426">426</a></p>
<p id="n427" class="pln"><a href="#n427">427</a></p>
<p id="n428" class="pln"><a href="#n428">428</a></p>
<p id="n429" class="pln"><a href="#n429">429</a></p>
<p id="n430" class="pln"><a href="#n430">430</a></p>
<p id="n431" class="pln"><a href="#n431">431</a></p>
<p id="n432" class="pln"><a href="#n432">432</a></p>
<p id="n433" class="stm mis"><a href="#n433">433</a></p>
<p id="n434" class="stm mis"><a href="#n434">434</a></p>
<p id="n435" class="stm mis"><a href="#n435">435</a></p>
<p id="n436" class="stm mis"><a href="#n436">436</a></p>
<p id="n437" class="stm mis"><a href="#n437">437</a></p>
<p id="n438" class="stm mis"><a href="#n438">438</a></p>
<p id="n439" class="stm mis"><a href="#n439">439</a></p>
<p id="n440" class="stm mis"><a href="#n440">440</a></p>
<p id="n441" class="stm mis"><a href="#n441">441</a></p>
<p id="n442" class="stm mis"><a href="#n442">442</a></p>
<p id="n443" class="stm mis"><a href="#n443">443</a></p>
<p id="n444" class="stm mis"><a href="#n444">444</a></p>
<p id="n445" class="stm mis"><a href="#n445">445</a></p>
<p id="n446" class="stm mis"><a href="#n446">446</a></p>
<p id="n447" class="stm mis"><a href="#n447">447</a></p>
<p id="n448" class="stm mis"><a href="#n448">448</a></p>
<p id="n449" class="stm mis"><a href="#n449">449</a></p>
<p id="n450" class="pln"><a href="#n450">450</a></p>
<p id="n451" class="pln"><a href="#n451">451</a></p>
<p id="n452" class="stm run hide_run"><a href="#n452">452</a></p>
<p id="n453" class="pln"><a href="#n453">453</a></p>
<p id="n454" class="pln"><a href="#n454">454</a></p>
<p id="n455" class="pln"><a href="#n455">455</a></p>
<p id="n456" class="pln"><a href="#n456">456</a></p>
<p id="n457" class="pln"><a href="#n457">457</a></p>
<p id="n458" class="pln"><a href="#n458">458</a></p>
<p id="n459" class="pln"><a href="#n459">459</a></p>
<p id="n460" class="pln"><a href="#n460">460</a></p>
<p id="n461" class="pln"><a href="#n461">461</a></p>
<p id="n462" class="pln"><a href="#n462">462</a></p>
<p id="n463" class="pln"><a href="#n463">463</a></p>
<p id="n464" class="pln"><a href="#n464">464</a></p>
<p id="n465" class="pln"><a href="#n465">465</a></p>
<p id="n466" class="pln"><a href="#n466">466</a></p>
<p id="n467" class="pln"><a href="#n467">467</a></p>
<p id="n468" class="stm mis"><a href="#n468">468</a></p>
<p id="n469" class="pln"><a href="#n469">469</a></p>
<p id="n470" class="stm mis"><a href="#n470">470</a></p>
<p id="n471" class="pln"><a href="#n471">471</a></p>
<p id="n472" class="pln"><a href="#n472">472</a></p>
<p id="n473" class="pln"><a href="#n473">473</a></p>
<p id="n474" class="stm mis"><a href="#n474">474</a></p>
<p id="n475" class="stm mis"><a href="#n475">475</a></p>
<p id="n476" class="pln"><a href="#n476">476</a></p>
<p id="n477" class="stm mis"><a href="#n477">477</a></p>
<p id="n478" class="pln"><a href="#n478">478</a></p>
<p id="n479" class="pln"><a href="#n479">479</a></p>
<p id="n480" class="stm run hide_run"><a href="#n480">480</a></p>
<p id="n481" class="stm mis"><a href="#n481">481</a></p>
<p id="n482" class="stm mis"><a href="#n482">482</a></p>

            </td>
            <td class="text">
<p id="t1" class="pln"><span class="com"># Copyright 2008-2018 by Peter Cock.  All rights reserved.</span><span class="strut">&nbsp;</span></p>
<p id="t2" class="pln"><span class="com">#</span><span class="strut">&nbsp;</span></p>
<p id="t3" class="pln"><span class="com"># This file is part of the Biopython distribution and governed by your</span><span class="strut">&nbsp;</span></p>
<p id="t4" class="pln"><span class="com"># choice of the "Biopython License Agreement" or the "BSD 3-Clause License".</span><span class="strut">&nbsp;</span></p>
<p id="t5" class="pln"><span class="com"># Please see the LICENSE file that should have been included as part of this</span><span class="strut">&nbsp;</span></p>
<p id="t6" class="pln"><span class="com"># package.</span><span class="strut">&nbsp;</span></p>
<p id="t7" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t8" class="stm run hide_run"><span class="str">"""Multiple sequence alignment input/output as alignment objects.</span><span class="strut">&nbsp;</span></p>
<p id="t9" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t10" class="pln"><span class="str">The Bio.AlignIO interface is deliberately very similar to Bio.SeqIO, and in</span><span class="strut">&nbsp;</span></p>
<p id="t11" class="pln"><span class="str">fact the two are connected internally.  Both modules use the same set of file</span><span class="strut">&nbsp;</span></p>
<p id="t12" class="pln"><span class="str">format names (lower case strings).  From the user's perspective, you can read</span><span class="strut">&nbsp;</span></p>
<p id="t13" class="pln"><span class="str">in a PHYLIP file containing one or more alignments using Bio.AlignIO, or you</span><span class="strut">&nbsp;</span></p>
<p id="t14" class="pln"><span class="str">can read in the sequences within these alignmenta using Bio.SeqIO.</span><span class="strut">&nbsp;</span></p>
<p id="t15" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t16" class="pln"><span class="str">Bio.AlignIO is also documented at http://biopython.org/wiki/AlignIO and by</span><span class="strut">&nbsp;</span></p>
<p id="t17" class="pln"><span class="str">a whole chapter in our tutorial:</span><span class="strut">&nbsp;</span></p>
<p id="t18" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t19" class="pln"><span class="str">* `HTML Tutorial`_</span><span class="strut">&nbsp;</span></p>
<p id="t20" class="pln"><span class="str">* `PDF Tutorial`_</span><span class="strut">&nbsp;</span></p>
<p id="t21" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t22" class="pln"><span class="str">.. _`HTML Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.html</span><span class="strut">&nbsp;</span></p>
<p id="t23" class="pln"><span class="str">.. _`PDF Tutorial`: http://biopython.org/DIST/docs/tutorial/Tutorial.pdf</span><span class="strut">&nbsp;</span></p>
<p id="t24" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t25" class="pln"><span class="str">Input</span><span class="strut">&nbsp;</span></p>
<p id="t26" class="pln"><span class="str">-----</span><span class="strut">&nbsp;</span></p>
<p id="t27" class="pln"><span class="str">For the typical special case when your file or handle contains one and only</span><span class="strut">&nbsp;</span></p>
<p id="t28" class="pln"><span class="str">one alignment, use the function Bio.AlignIO.read().  This takes an input file</span><span class="strut">&nbsp;</span></p>
<p id="t29" class="pln"><span class="str">handle (or in recent versions of Biopython a filename as a string), format</span><span class="strut">&nbsp;</span></p>
<p id="t30" class="pln"><span class="str">string and optional number of sequences per alignment.  It will return a single</span><span class="strut">&nbsp;</span></p>
<p id="t31" class="pln"><span class="str">MultipleSeqAlignment object (or raise an exception if there isn't just one</span><span class="strut">&nbsp;</span></p>
<p id="t32" class="pln"><span class="str">alignment):</span><span class="strut">&nbsp;</span></p>
<p id="t33" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t34" class="pln"><span class="str">>>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t35" class="pln"><span class="str">>>> align = AlignIO.read("Phylip/interlaced.phy", "phylip")</span><span class="strut">&nbsp;</span></p>
<p id="t36" class="pln"><span class="str">>>> print(align)</span><span class="strut">&nbsp;</span></p>
<p id="t37" class="pln"><span class="str">SingleLetterAlphabet() alignment with 3 rows and 384 columns</span><span class="strut">&nbsp;</span></p>
<p id="t38" class="pln"><span class="str">-----MKVILLFVLAVFTVFVSS---------------RGIPPE...I-- CYS1_DICDI</span><span class="strut">&nbsp;</span></p>
<p id="t39" class="pln"><span class="str">MAHARVLLLALAVLATAAVAVASSSSFADSNPIRPVTDRAASTL...VAA ALEU_HORVU</span><span class="strut">&nbsp;</span></p>
<p id="t40" class="pln"><span class="str">------MWATLPLLCAGAWLLGV--------PVCGAAELSVNSL...PLV CATH_HUMAN</span><span class="strut">&nbsp;</span></p>
<p id="t41" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t42" class="pln"><span class="str">For the general case, when the handle could contain any number of alignments,</span><span class="strut">&nbsp;</span></p>
<p id="t43" class="pln"><span class="str">use the function Bio.AlignIO.parse(...) which takes the same arguments, but</span><span class="strut">&nbsp;</span></p>
<p id="t44" class="pln"><span class="str">returns an iterator giving MultipleSeqAlignment objects (typically used in a</span><span class="strut">&nbsp;</span></p>
<p id="t45" class="pln"><span class="str">for loop). If you want random access to the alignments by number, turn this</span><span class="strut">&nbsp;</span></p>
<p id="t46" class="pln"><span class="str">into a list:</span><span class="strut">&nbsp;</span></p>
<p id="t47" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t48" class="pln"><span class="str">>>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t49" class="pln"><span class="str">>>> alignments = list(AlignIO.parse("Emboss/needle.txt", "emboss"))</span><span class="strut">&nbsp;</span></p>
<p id="t50" class="pln"><span class="str">>>> print(alignments[2])</span><span class="strut">&nbsp;</span></p>
<p id="t51" class="pln"><span class="str">SingleLetterAlphabet() alignment with 2 rows and 120 columns</span><span class="strut">&nbsp;</span></p>
<p id="t52" class="pln"><span class="str">-KILIVDDQYGIRILLNEVFNKEGYQTFQAANGLQALDIVTKER...--- ref_rec</span><span class="strut">&nbsp;</span></p>
<p id="t53" class="pln"><span class="str">LHIVVVDDDPGTCVYIESVFAELGHTCKSFVRPEAAEEYILTHP...HKE gi|94967506|receiver</span><span class="strut">&nbsp;</span></p>
<p id="t54" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t55" class="pln"><span class="str">Most alignment file formats can be concatenated so as to hold as many</span><span class="strut">&nbsp;</span></p>
<p id="t56" class="pln"><span class="str">different multiple sequence alignments as possible.  One common example</span><span class="strut">&nbsp;</span></p>
<p id="t57" class="pln"><span class="str">is the output of the tool seqboot in the PHLYIP suite.  Sometimes there</span><span class="strut">&nbsp;</span></p>
<p id="t58" class="pln"><span class="str">can be a file header and footer, as seen in the EMBOSS alignment output.</span><span class="strut">&nbsp;</span></p>
<p id="t59" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t60" class="pln"><span class="str">Output</span><span class="strut">&nbsp;</span></p>
<p id="t61" class="pln"><span class="str">------</span><span class="strut">&nbsp;</span></p>
<p id="t62" class="pln"><span class="str">Use the function Bio.AlignIO.write(...), which takes a complete set of</span><span class="strut">&nbsp;</span></p>
<p id="t63" class="pln"><span class="str">Alignment objects (either as a list, or an iterator), an output file handle</span><span class="strut">&nbsp;</span></p>
<p id="t64" class="pln"><span class="str">(or filename in recent versions of Biopython) and of course the file format::</span><span class="strut">&nbsp;</span></p>
<p id="t65" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t66" class="pln"><span class="str">    from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t67" class="pln"><span class="str">    alignments = ...</span><span class="strut">&nbsp;</span></p>
<p id="t68" class="pln"><span class="str">    count = SeqIO.write(alignments, "example.faa", "fasta")</span><span class="strut">&nbsp;</span></p>
<p id="t69" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t70" class="pln"><span class="str">If using a handle make sure to close it to flush the data to the disk::</span><span class="strut">&nbsp;</span></p>
<p id="t71" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t72" class="pln"><span class="str">    from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t73" class="pln"><span class="str">    alignments = ...</span><span class="strut">&nbsp;</span></p>
<p id="t74" class="pln"><span class="str">    with open("example.faa", "w") as handle:</span><span class="strut">&nbsp;</span></p>
<p id="t75" class="pln"><span class="str">        count = SeqIO.write(alignments, handle, "fasta")</span><span class="strut">&nbsp;</span></p>
<p id="t76" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t77" class="pln"><span class="str">In general, you are expected to call this function once (with all your</span><span class="strut">&nbsp;</span></p>
<p id="t78" class="pln"><span class="str">alignments) and then close the file handle.  However, for file formats</span><span class="strut">&nbsp;</span></p>
<p id="t79" class="pln"><span class="str">like PHYLIP where multiple alignments are stored sequentially (with no file</span><span class="strut">&nbsp;</span></p>
<p id="t80" class="pln"><span class="str">header and footer), then multiple calls to the write function should work as</span><span class="strut">&nbsp;</span></p>
<p id="t81" class="pln"><span class="str">expected when using handles.</span><span class="strut">&nbsp;</span></p>
<p id="t82" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t83" class="pln"><span class="str">If you are using a filename, the repeated calls to the write functions will</span><span class="strut">&nbsp;</span></p>
<p id="t84" class="pln"><span class="str">overwrite the existing file each time.</span><span class="strut">&nbsp;</span></p>
<p id="t85" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t86" class="pln"><span class="str">Conversion</span><span class="strut">&nbsp;</span></p>
<p id="t87" class="pln"><span class="str">----------</span><span class="strut">&nbsp;</span></p>
<p id="t88" class="pln"><span class="str">The Bio.AlignIO.convert(...) function allows an easy interface for simple</span><span class="strut">&nbsp;</span></p>
<p id="t89" class="pln"><span class="str">alignment file format conversions. Additionally, it may use file format</span><span class="strut">&nbsp;</span></p>
<p id="t90" class="pln"><span class="str">specific optimisations so this should be the fastest way too.</span><span class="strut">&nbsp;</span></p>
<p id="t91" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t92" class="pln"><span class="str">In general however, you can combine the Bio.AlignIO.parse(...) function with</span><span class="strut">&nbsp;</span></p>
<p id="t93" class="pln"><span class="str">the Bio.AlignIO.write(...) function for sequence file conversion. Using</span><span class="strut">&nbsp;</span></p>
<p id="t94" class="pln"><span class="str">generator expressions provides a memory efficient way to perform filtering or</span><span class="strut">&nbsp;</span></p>
<p id="t95" class="pln"><span class="str">other extra operations as part of the process.</span><span class="strut">&nbsp;</span></p>
<p id="t96" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t97" class="pln"><span class="str">File Formats</span><span class="strut">&nbsp;</span></p>
<p id="t98" class="pln"><span class="str">------------</span><span class="strut">&nbsp;</span></p>
<p id="t99" class="pln"><span class="str">When specifying the file format, use lowercase strings.  The same format</span><span class="strut">&nbsp;</span></p>
<p id="t100" class="pln"><span class="str">names are also used in Bio.SeqIO and include the following:</span><span class="strut">&nbsp;</span></p>
<p id="t101" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t102" class="pln"><span class="str">  - clustal -   Output from Clustal W or X, see also the module Bio.Clustalw</span><span class="strut">&nbsp;</span></p>
<p id="t103" class="pln"><span class="str">    which can be used to run the command line tool from Biopython.</span><span class="strut">&nbsp;</span></p>
<p id="t104" class="pln"><span class="str">  - emboss    - EMBOSS tools' "pairs" and "simple" alignment formats.</span><span class="strut">&nbsp;</span></p>
<p id="t105" class="pln"><span class="str">  - fasta     - The generic sequence file format where each record starts with</span><span class="strut">&nbsp;</span></p>
<p id="t106" class="pln"><span class="str">    an identifer line starting with a ">" character, followed by</span><span class="strut">&nbsp;</span></p>
<p id="t107" class="pln"><span class="str">    lines of sequence.</span><span class="strut">&nbsp;</span></p>
<p id="t108" class="pln"><span class="str">  - fasta-m10 - For the pairswise alignments output by Bill Pearson's FASTA</span><span class="strut">&nbsp;</span></p>
<p id="t109" class="pln"><span class="str">    tools when used with the -m 10 command line option for machine</span><span class="strut">&nbsp;</span></p>
<p id="t110" class="pln"><span class="str">    readable output.</span><span class="strut">&nbsp;</span></p>
<p id="t111" class="pln"><span class="str">  - ig        - The IntelliGenetics file format, apparently the same as the</span><span class="strut">&nbsp;</span></p>
<p id="t112" class="pln"><span class="str">    MASE alignment format.</span><span class="strut">&nbsp;</span></p>
<p id="t113" class="pln"><span class="str">  - nexus     - Output from NEXUS, see also the module Bio.Nexus which can also</span><span class="strut">&nbsp;</span></p>
<p id="t114" class="pln"><span class="str">    read any phylogenetic trees in these files.</span><span class="strut">&nbsp;</span></p>
<p id="t115" class="pln"><span class="str">  - phylip    - Interlaced PHYLIP, as used by the PHLIP tools.</span><span class="strut">&nbsp;</span></p>
<p id="t116" class="pln"><span class="str">  - phylip-sequential - Sequential PHYLIP.</span><span class="strut">&nbsp;</span></p>
<p id="t117" class="pln"><span class="str">  - phylip-relaxed - PHYLIP like format allowing longer names.</span><span class="strut">&nbsp;</span></p>
<p id="t118" class="pln"><span class="str">  - stockholm - A richly annotated alignment file format used by PFAM.</span><span class="strut">&nbsp;</span></p>
<p id="t119" class="pln"><span class="str">  - mauve - Output from progressiveMauve/Mauve</span><span class="strut">&nbsp;</span></p>
<p id="t120" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t121" class="pln"><span class="str">Note that while Bio.AlignIO can read all the above file formats, it cannot</span><span class="strut">&nbsp;</span></p>
<p id="t122" class="pln"><span class="str">write to all of them.</span><span class="strut">&nbsp;</span></p>
<p id="t123" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t124" class="pln"><span class="str">You can also use any file format supported by Bio.SeqIO, such as "fasta" or</span><span class="strut">&nbsp;</span></p>
<p id="t125" class="pln"><span class="str">"ig" (which are listed above), PROVIDED the sequences in your file are all the</span><span class="strut">&nbsp;</span></p>
<p id="t126" class="pln"><span class="str">same length.</span><span class="strut">&nbsp;</span></p>
<p id="t127" class="pln"><span class="str">"""</span><span class="strut">&nbsp;</span></p>
<p id="t128" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t129" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t130" class="stm run hide_run"><span class="key">from</span> <span class="nam">__future__</span> <span class="key">import</span> <span class="nam">print_function</span><span class="strut">&nbsp;</span></p>
<p id="t131" class="stm run hide_run"><span class="key">from</span> <span class="nam">Bio</span><span class="op">.</span><span class="nam">_py3k</span> <span class="key">import</span> <span class="nam">basestring</span><span class="strut">&nbsp;</span></p>
<p id="t132" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t133" class="pln"><span class="com"># TODO</span><span class="strut">&nbsp;</span></p>
<p id="t134" class="pln"><span class="com"># - define policy on reading aligned sequences with gaps in</span><span class="strut">&nbsp;</span></p>
<p id="t135" class="pln"><span class="com">#   (e.g. - and . characters) including how the alphabet interacts</span><span class="strut">&nbsp;</span></p>
<p id="t136" class="pln"><span class="com">#</span><span class="strut">&nbsp;</span></p>
<p id="t137" class="pln"><span class="com"># - Can we build the to_alignment(...) functionality</span><span class="strut">&nbsp;</span></p>
<p id="t138" class="pln"><span class="com">#   into the generic Alignment class instead?</span><span class="strut">&nbsp;</span></p>
<p id="t139" class="pln"><span class="com">#</span><span class="strut">&nbsp;</span></p>
<p id="t140" class="pln"><span class="com"># - How best to handle unique/non unique record.id when writing.</span><span class="strut">&nbsp;</span></p>
<p id="t141" class="pln"><span class="com">#   For most file formats reading such files is fine; The stockholm</span><span class="strut">&nbsp;</span></p>
<p id="t142" class="pln"><span class="com">#   parser would fail.</span><span class="strut">&nbsp;</span></p>
<p id="t143" class="pln"><span class="com">#</span><span class="strut">&nbsp;</span></p>
<p id="t144" class="pln"><span class="com"># - MSF multiple alignment format, aka GCG, aka PileUp format (*.msf)</span><span class="strut">&nbsp;</span></p>
<p id="t145" class="pln"><span class="com">#   http://www.bioperl.org/wiki/MSF_multiple_alignment_format</span><span class="strut">&nbsp;</span></p>
<p id="t146" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t147" class="stm run hide_run"><span class="key">from</span> <span class="nam">Bio</span><span class="op">.</span><span class="nam">Align</span> <span class="key">import</span> <span class="nam">MultipleSeqAlignment</span><span class="strut">&nbsp;</span></p>
<p id="t148" class="stm run hide_run"><span class="key">from</span> <span class="nam">Bio</span><span class="op">.</span><span class="nam">Alphabet</span> <span class="key">import</span> <span class="nam">Alphabet</span><span class="op">,</span> <span class="nam">AlphabetEncoder</span><span class="op">,</span> <span class="nam">_get_base_alphabet</span><span class="strut">&nbsp;</span></p>
<p id="t149" class="stm run hide_run"><span class="key">from</span> <span class="nam">Bio</span><span class="op">.</span><span class="nam">File</span> <span class="key">import</span> <span class="nam">as_handle</span><span class="strut">&nbsp;</span></p>
<p id="t150" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t151" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">StockholmIO</span><span class="strut">&nbsp;</span></p>
<p id="t152" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">ClustalIO</span><span class="strut">&nbsp;</span></p>
<p id="t153" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">NexusIO</span><span class="strut">&nbsp;</span></p>
<p id="t154" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">PhylipIO</span><span class="strut">&nbsp;</span></p>
<p id="t155" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">EmbossIO</span><span class="strut">&nbsp;</span></p>
<p id="t156" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">FastaIO</span><span class="strut">&nbsp;</span></p>
<p id="t157" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">MafIO</span><span class="strut">&nbsp;</span></p>
<p id="t158" class="stm run hide_run"><span class="key">from</span> <span class="op">.</span> <span class="key">import</span> <span class="nam">MauveIO</span><span class="strut">&nbsp;</span></p>
<p id="t159" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t160" class="pln"><span class="com"># Convention for format names is "mainname-subtype" in lower case.</span><span class="strut">&nbsp;</span></p>
<p id="t161" class="pln"><span class="com"># Please use the same names as BioPerl and EMBOSS where possible.</span><span class="strut">&nbsp;</span></p>
<p id="t162" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t163" class="stm run hide_run"><span class="nam">_FormatToIterator</span> <span class="op">=</span> <span class="op">{</span>  <span class="com"># "fasta" is done via Bio.SeqIO</span><span class="strut">&nbsp;</span></p>
<p id="t164" class="pln">                     <span class="str">"clustal"</span><span class="op">:</span> <span class="nam">ClustalIO</span><span class="op">.</span><span class="nam">ClustalIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t165" class="pln">                     <span class="str">"emboss"</span><span class="op">:</span> <span class="nam">EmbossIO</span><span class="op">.</span><span class="nam">EmbossIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t166" class="pln">                     <span class="str">"fasta-m10"</span><span class="op">:</span> <span class="nam">FastaIO</span><span class="op">.</span><span class="nam">FastaM10Iterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t167" class="pln">                     <span class="str">"maf"</span><span class="op">:</span> <span class="nam">MafIO</span><span class="op">.</span><span class="nam">MafIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t168" class="pln">                     <span class="str">"mauve"</span><span class="op">:</span> <span class="nam">MauveIO</span><span class="op">.</span><span class="nam">MauveIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t169" class="pln">                     <span class="str">"nexus"</span><span class="op">:</span> <span class="nam">NexusIO</span><span class="op">.</span><span class="nam">NexusIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t170" class="pln">                     <span class="str">"phylip"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">PhylipIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t171" class="pln">                     <span class="str">"phylip-sequential"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">SequentialPhylipIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t172" class="pln">                     <span class="str">"phylip-relaxed"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">RelaxedPhylipIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t173" class="pln">                     <span class="str">"stockholm"</span><span class="op">:</span> <span class="nam">StockholmIO</span><span class="op">.</span><span class="nam">StockholmIterator</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t174" class="pln">                     <span class="op">}</span><span class="strut">&nbsp;</span></p>
<p id="t175" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t176" class="stm run hide_run"><span class="nam">_FormatToWriter</span> <span class="op">=</span> <span class="op">{</span>  <span class="com"># "fasta" is done via Bio.SeqIO</span><span class="strut">&nbsp;</span></p>
<p id="t177" class="pln">                     <span class="com"># "emboss" : EmbossIO.EmbossWriter, (unfinished)</span><span class="strut">&nbsp;</span></p>
<p id="t178" class="pln">                   <span class="str">"clustal"</span><span class="op">:</span> <span class="nam">ClustalIO</span><span class="op">.</span><span class="nam">ClustalWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t179" class="pln">                   <span class="str">"maf"</span><span class="op">:</span> <span class="nam">MafIO</span><span class="op">.</span><span class="nam">MafWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t180" class="pln">                   <span class="str">"mauve"</span><span class="op">:</span> <span class="nam">MauveIO</span><span class="op">.</span><span class="nam">MauveWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t181" class="pln">                   <span class="str">"nexus"</span><span class="op">:</span> <span class="nam">NexusIO</span><span class="op">.</span><span class="nam">NexusWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t182" class="pln">                   <span class="str">"phylip"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">PhylipWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t183" class="pln">                   <span class="str">"phylip-sequential"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">SequentialPhylipWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t184" class="pln">                   <span class="str">"phylip-relaxed"</span><span class="op">:</span> <span class="nam">PhylipIO</span><span class="op">.</span><span class="nam">RelaxedPhylipWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t185" class="pln">                   <span class="str">"stockholm"</span><span class="op">:</span> <span class="nam">StockholmIO</span><span class="op">.</span><span class="nam">StockholmWriter</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t186" class="pln">                   <span class="op">}</span><span class="strut">&nbsp;</span></p>
<p id="t187" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t188" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t189" class="stm run hide_run"><span class="key">def</span> <span class="nam">write</span><span class="op">(</span><span class="nam">alignments</span><span class="op">,</span> <span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t190" class="pln">    <span class="str">"""Write complete set of alignments to a file.</span><span class="strut">&nbsp;</span></p>
<p id="t191" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t192" class="pln"><span class="str">    Arguments:</span><span class="strut">&nbsp;</span></p>
<p id="t193" class="pln"><span class="str">     - alignments - A list (or iterator) of MultipleSeqAlignment objects,</span><span class="strut">&nbsp;</span></p>
<p id="t194" class="pln"><span class="str">       or a single alignment object.</span><span class="strut">&nbsp;</span></p>
<p id="t195" class="pln"><span class="str">     - handle    - File handle object to write to, or filename as string</span><span class="strut">&nbsp;</span></p>
<p id="t196" class="pln"><span class="str">       (note older versions of Biopython only took a handle).</span><span class="strut">&nbsp;</span></p>
<p id="t197" class="pln"><span class="str">     - format    - lower case string describing the file format to write.</span><span class="strut">&nbsp;</span></p>
<p id="t198" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t199" class="pln"><span class="str">    You should close the handle after calling this function.</span><span class="strut">&nbsp;</span></p>
<p id="t200" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t201" class="pln"><span class="str">    Returns the number of alignments written (as an integer).</span><span class="strut">&nbsp;</span></p>
<p id="t202" class="pln"><span class="str">    """</span><span class="strut">&nbsp;</span></p>
<p id="t203" class="stm mis">    <span class="key">from</span> <span class="nam">Bio</span> <span class="key">import</span> <span class="nam">SeqIO</span><span class="strut">&nbsp;</span></p>
<p id="t204" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t205" class="pln">    <span class="com"># Try and give helpful error messages:</span><span class="strut">&nbsp;</span></p>
<p id="t206" class="stm mis">    <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">format</span><span class="op">,</span> <span class="nam">basestring</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t207" class="stm mis">        <span class="key">raise</span> <span class="nam">TypeError</span><span class="op">(</span><span class="str">"Need a string for the file format (lower case)"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t208" class="stm mis">    <span class="key">if</span> <span class="key">not</span> <span class="nam">format</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t209" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Format required (lower case string)"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t210" class="stm mis">    <span class="key">if</span> <span class="nam">format</span> <span class="op">!=</span> <span class="nam">format</span><span class="op">.</span><span class="nam">lower</span><span class="op">(</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t211" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Format string '%s' should be lower case"</span> <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t212" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t213" class="stm mis">    <span class="key">if</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">alignments</span><span class="op">,</span> <span class="nam">MultipleSeqAlignment</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t214" class="pln">        <span class="com"># This raised an exception in older versions of Biopython</span><span class="strut">&nbsp;</span></p>
<p id="t215" class="stm mis">        <span class="nam">alignments</span> <span class="op">=</span> <span class="op">[</span><span class="nam">alignments</span><span class="op">]</span><span class="strut">&nbsp;</span></p>
<p id="t216" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t217" class="stm mis">    <span class="key">with</span> <span class="nam">as_handle</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="str">'w'</span><span class="op">)</span> <span class="key">as</span> <span class="nam">fp</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t218" class="pln">        <span class="com"># Map the file format to a writer class</span><span class="strut">&nbsp;</span></p>
<p id="t219" class="stm mis">        <span class="key">if</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">_FormatToWriter</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t220" class="stm mis">            <span class="nam">writer_class</span> <span class="op">=</span> <span class="nam">_FormatToWriter</span><span class="op">[</span><span class="nam">format</span><span class="op">]</span><span class="strut">&nbsp;</span></p>
<p id="t221" class="stm mis">            <span class="nam">count</span> <span class="op">=</span> <span class="nam">writer_class</span><span class="op">(</span><span class="nam">fp</span><span class="op">)</span><span class="op">.</span><span class="nam">write_file</span><span class="op">(</span><span class="nam">alignments</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t222" class="stm mis">        <span class="key">elif</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">_FormatToWriter</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t223" class="pln">            <span class="com"># Exploit the existing SeqIO parser to do the dirty work!</span><span class="strut">&nbsp;</span></p>
<p id="t224" class="pln">            <span class="com"># TODO - Can we make one call to SeqIO.write() and count the alignments?</span><span class="strut">&nbsp;</span></p>
<p id="t225" class="stm mis">            <span class="nam">count</span> <span class="op">=</span> <span class="num">0</span><span class="strut">&nbsp;</span></p>
<p id="t226" class="stm mis">            <span class="key">for</span> <span class="nam">alignment</span> <span class="key">in</span> <span class="nam">alignments</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t227" class="stm mis">                <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">alignment</span><span class="op">,</span> <span class="nam">MultipleSeqAlignment</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t228" class="stm mis">                    <span class="key">raise</span> <span class="nam">TypeError</span><span class="op">(</span><span class="str">"Expect a list or iterator of MultipleSeqAlignment "</span><span class="strut">&nbsp;</span></p>
<p id="t229" class="pln">                                    <span class="str">"objects, got: %r"</span> <span class="op">%</span> <span class="nam">alignment</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t230" class="stm mis">                <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">write</span><span class="op">(</span><span class="nam">alignment</span><span class="op">,</span> <span class="nam">fp</span><span class="op">,</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t231" class="stm mis">                <span class="nam">count</span> <span class="op">+=</span> <span class="num">1</span><span class="strut">&nbsp;</span></p>
<p id="t232" class="stm mis">        <span class="key">elif</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">_FormatToIterator</span> <span class="key">or</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">_FormatToIterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t233" class="stm mis">            <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Reading format '%s' is supported, but not writing"</span><span class="strut">&nbsp;</span></p>
<p id="t234" class="pln">                             <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t235" class="pln">        <span class="key">else</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t236" class="stm mis">            <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Unknown format '%s'"</span> <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t237" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t238" class="stm mis">    <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">count</span><span class="op">,</span> <span class="nam">int</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t239" class="stm mis">        <span class="key">raise</span> <span class="nam">RuntimeError</span><span class="op">(</span><span class="str">"Internal error - the underlying %s "</span><span class="strut">&nbsp;</span></p>
<p id="t240" class="pln">                           <span class="str">"writer should have returned the alignment count, not %s"</span><span class="strut">&nbsp;</span></p>
<p id="t241" class="pln">                           <span class="op">%</span> <span class="op">(</span><span class="nam">format</span><span class="op">,</span> <span class="nam">repr</span><span class="op">(</span><span class="nam">count</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t242" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t243" class="stm mis">    <span class="key">return</span> <span class="nam">count</span><span class="strut">&nbsp;</span></p>
<p id="t244" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t245" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t246" class="pln"><span class="com"># This is a generator function!</span><span class="strut">&nbsp;</span></p>
<p id="t247" class="stm run hide_run"><span class="key">def</span> <span class="nam">_SeqIO_to_alignment_iterator</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">=</span><span class="key">None</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">=</span><span class="key">None</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t248" class="pln">    <span class="str">"""Use Bio.SeqIO to create an MultipleSeqAlignment iterator (PRIVATE).</span><span class="strut">&nbsp;</span></p>
<p id="t249" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t250" class="pln"><span class="str">    Arguments:</span><span class="strut">&nbsp;</span></p>
<p id="t251" class="pln"><span class="str">     - handle    - handle to the file.</span><span class="strut">&nbsp;</span></p>
<p id="t252" class="pln"><span class="str">     - format    - string describing the file format.</span><span class="strut">&nbsp;</span></p>
<p id="t253" class="pln"><span class="str">     - alphabet  - optional Alphabet object, useful when the sequence type</span><span class="strut">&nbsp;</span></p>
<p id="t254" class="pln"><span class="str">       cannot be automatically inferred from the file itself</span><span class="strut">&nbsp;</span></p>
<p id="t255" class="pln"><span class="str">       (e.g. fasta, phylip, clustal)</span><span class="strut">&nbsp;</span></p>
<p id="t256" class="pln"><span class="str">     - seq_count - Optional integer, number of sequences expected in each</span><span class="strut">&nbsp;</span></p>
<p id="t257" class="pln"><span class="str">       alignment.  Recommended for fasta format files.</span><span class="strut">&nbsp;</span></p>
<p id="t258" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t259" class="pln"><span class="str">    If count is omitted (default) then all the sequences in the file are</span><span class="strut">&nbsp;</span></p>
<p id="t260" class="pln"><span class="str">    combined into a single MultipleSeqAlignment.</span><span class="strut">&nbsp;</span></p>
<p id="t261" class="pln"><span class="str">    """</span><span class="strut">&nbsp;</span></p>
<p id="t262" class="stm mis">    <span class="key">from</span> <span class="nam">Bio</span> <span class="key">import</span> <span class="nam">SeqIO</span><span class="strut">&nbsp;</span></p>
<p id="t263" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t264" class="stm mis">    <span class="key">if</span> <span class="nam">format</span> <span class="key">not</span> <span class="key">in</span> <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">_FormatToIterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t265" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Unknown format '%s'"</span> <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t266" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t267" class="stm mis">    <span class="key">if</span> <span class="nam">seq_count</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t268" class="pln">        <span class="com"># Use the count to split the records into batches.</span><span class="strut">&nbsp;</span></p>
<p id="t269" class="stm mis">        <span class="nam">seq_record_iterator</span> <span class="op">=</span> <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">parse</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t270" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t271" class="stm mis">        <span class="nam">records</span> <span class="op">=</span> <span class="op">[</span><span class="op">]</span><span class="strut">&nbsp;</span></p>
<p id="t272" class="stm mis">        <span class="key">for</span> <span class="nam">record</span> <span class="key">in</span> <span class="nam">seq_record_iterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t273" class="stm mis">            <span class="nam">records</span><span class="op">.</span><span class="nam">append</span><span class="op">(</span><span class="nam">record</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t274" class="stm mis">            <span class="key">if</span> <span class="nam">len</span><span class="op">(</span><span class="nam">records</span><span class="op">)</span> <span class="op">==</span> <span class="nam">seq_count</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t275" class="stm mis">                <span class="key">yield</span> <span class="nam">MultipleSeqAlignment</span><span class="op">(</span><span class="nam">records</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t276" class="stm mis">                <span class="nam">records</span> <span class="op">=</span> <span class="op">[</span><span class="op">]</span><span class="strut">&nbsp;</span></p>
<p id="t277" class="stm mis">        <span class="key">if</span> <span class="nam">records</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t278" class="stm mis">            <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Check seq_count argument, not enough sequences?"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t279" class="pln">    <span class="key">else</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t280" class="pln">        <span class="com"># Must assume that there is a single alignment using all</span><span class="strut">&nbsp;</span></p>
<p id="t281" class="pln">        <span class="com"># the SeqRecord objects:</span><span class="strut">&nbsp;</span></p>
<p id="t282" class="stm mis">        <span class="nam">records</span> <span class="op">=</span> <span class="nam">list</span><span class="op">(</span><span class="nam">SeqIO</span><span class="op">.</span><span class="nam">parse</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t283" class="stm mis">        <span class="key">if</span> <span class="nam">records</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t284" class="stm mis">            <span class="key">yield</span> <span class="nam">MultipleSeqAlignment</span><span class="op">(</span><span class="nam">records</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t285" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t286" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t287" class="stm run hide_run"><span class="key">def</span> <span class="nam">_force_alphabet</span><span class="op">(</span><span class="nam">alignment_iterator</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t288" class="pln">    <span class="str">"""Iterate over alignments, over-riding the alphabet (PRIVATE)."""</span><span class="strut">&nbsp;</span></p>
<p id="t289" class="pln">    <span class="com"># Assume the alphabet argument has been pre-validated</span><span class="strut">&nbsp;</span></p>
<p id="t290" class="stm mis">    <span class="nam">given_base_class</span> <span class="op">=</span> <span class="nam">_get_base_alphabet</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">.</span><span class="nam">__class__</span><span class="strut">&nbsp;</span></p>
<p id="t291" class="stm mis">    <span class="key">for</span> <span class="nam">align</span> <span class="key">in</span> <span class="nam">alignment_iterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t292" class="stm mis">        <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">_get_base_alphabet</span><span class="op">(</span><span class="nam">align</span><span class="op">.</span><span class="nam">_alphabet</span><span class="op">)</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t293" class="pln">                          <span class="nam">given_base_class</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t294" class="stm mis">            <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Specified alphabet %s clashes with "</span><span class="strut">&nbsp;</span></p>
<p id="t295" class="pln">                             <span class="str">"that determined from the file, %s"</span><span class="strut">&nbsp;</span></p>
<p id="t296" class="pln">                             <span class="op">%</span> <span class="op">(</span><span class="nam">repr</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">,</span> <span class="nam">repr</span><span class="op">(</span><span class="nam">align</span><span class="op">.</span><span class="nam">_alphabet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t297" class="stm mis">        <span class="key">for</span> <span class="nam">record</span> <span class="key">in</span> <span class="nam">align</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t298" class="stm mis">            <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">_get_base_alphabet</span><span class="op">(</span><span class="nam">record</span><span class="op">.</span><span class="nam">seq</span><span class="op">.</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t299" class="pln">                              <span class="nam">given_base_class</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t300" class="stm mis">                <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Specified alphabet %s clashes with "</span><span class="strut">&nbsp;</span></p>
<p id="t301" class="pln">                                 <span class="str">"that determined from the file, %s"</span><span class="strut">&nbsp;</span></p>
<p id="t302" class="pln">                           <span class="op">%</span> <span class="op">(</span><span class="nam">repr</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">,</span> <span class="nam">repr</span><span class="op">(</span><span class="nam">record</span><span class="op">.</span><span class="nam">seq</span><span class="op">.</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t303" class="stm mis">            <span class="nam">record</span><span class="op">.</span><span class="nam">seq</span><span class="op">.</span><span class="nam">alphabet</span> <span class="op">=</span> <span class="nam">alphabet</span><span class="strut">&nbsp;</span></p>
<p id="t304" class="stm mis">        <span class="nam">align</span><span class="op">.</span><span class="nam">_alphabet</span> <span class="op">=</span> <span class="nam">alphabet</span><span class="strut">&nbsp;</span></p>
<p id="t305" class="stm mis">        <span class="key">yield</span> <span class="nam">align</span><span class="strut">&nbsp;</span></p>
<p id="t306" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t307" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t308" class="stm run hide_run"><span class="key">def</span> <span class="nam">parse</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">=</span><span class="key">None</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">=</span><span class="key">None</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t309" class="pln">    <span class="str">"""Iterate over an alignment file as MultipleSeqAlignment objects.</span><span class="strut">&nbsp;</span></p>
<p id="t310" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t311" class="pln"><span class="str">    Arguments:</span><span class="strut">&nbsp;</span></p>
<p id="t312" class="pln"><span class="str">     - handle    - handle to the file, or the filename as a string</span><span class="strut">&nbsp;</span></p>
<p id="t313" class="pln"><span class="str">       (note older versions of Biopython only took a handle).</span><span class="strut">&nbsp;</span></p>
<p id="t314" class="pln"><span class="str">     - format    - string describing the file format.</span><span class="strut">&nbsp;</span></p>
<p id="t315" class="pln"><span class="str">     - alphabet  - optional Alphabet object, useful when the sequence type</span><span class="strut">&nbsp;</span></p>
<p id="t316" class="pln"><span class="str">       cannot be automatically inferred from the file itself</span><span class="strut">&nbsp;</span></p>
<p id="t317" class="pln"><span class="str">       (e.g. fasta, phylip, clustal)</span><span class="strut">&nbsp;</span></p>
<p id="t318" class="pln"><span class="str">     - seq_count - Optional integer, number of sequences expected in each</span><span class="strut">&nbsp;</span></p>
<p id="t319" class="pln"><span class="str">       alignment.  Recommended for fasta format files.</span><span class="strut">&nbsp;</span></p>
<p id="t320" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t321" class="pln"><span class="str">    If you have the file name in a string 'filename', use:</span><span class="strut">&nbsp;</span></p>
<p id="t322" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t323" class="pln"><span class="str">    >>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t324" class="pln"><span class="str">    >>> filename = "Emboss/needle.txt"</span><span class="strut">&nbsp;</span></p>
<p id="t325" class="pln"><span class="str">    >>> format = "emboss"</span><span class="strut">&nbsp;</span></p>
<p id="t326" class="pln"><span class="str">    >>> for alignment in AlignIO.parse(filename, format):</span><span class="strut">&nbsp;</span></p>
<p id="t327" class="pln"><span class="str">    ...     print("Alignment of length %i" % alignment.get_alignment_length())</span><span class="strut">&nbsp;</span></p>
<p id="t328" class="pln"><span class="str">    Alignment of length 124</span><span class="strut">&nbsp;</span></p>
<p id="t329" class="pln"><span class="str">    Alignment of length 119</span><span class="strut">&nbsp;</span></p>
<p id="t330" class="pln"><span class="str">    Alignment of length 120</span><span class="strut">&nbsp;</span></p>
<p id="t331" class="pln"><span class="str">    Alignment of length 118</span><span class="strut">&nbsp;</span></p>
<p id="t332" class="pln"><span class="str">    Alignment of length 125</span><span class="strut">&nbsp;</span></p>
<p id="t333" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t334" class="pln"><span class="str">    If you have a string 'data' containing the file contents, use::</span><span class="strut">&nbsp;</span></p>
<p id="t335" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t336" class="pln"><span class="str">      from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t337" class="pln"><span class="str">      from StringIO import StringIO</span><span class="strut">&nbsp;</span></p>
<p id="t338" class="pln"><span class="str">      my_iterator = AlignIO.parse(StringIO(data), format)</span><span class="strut">&nbsp;</span></p>
<p id="t339" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t340" class="pln"><span class="str">    Use the Bio.AlignIO.read() function when you expect a single record only.</span><span class="strut">&nbsp;</span></p>
<p id="t341" class="pln"><span class="str">    """</span><span class="strut">&nbsp;</span></p>
<p id="t342" class="stm mis">    <span class="key">from</span> <span class="nam">Bio</span> <span class="key">import</span> <span class="nam">SeqIO</span><span class="strut">&nbsp;</span></p>
<p id="t343" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t344" class="pln">    <span class="com"># Try and give helpful error messages:</span><span class="strut">&nbsp;</span></p>
<p id="t345" class="stm mis">    <span class="key">if</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">format</span><span class="op">,</span> <span class="nam">basestring</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t346" class="stm mis">        <span class="key">raise</span> <span class="nam">TypeError</span><span class="op">(</span><span class="str">"Need a string for the file format (lower case)"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t347" class="stm mis">    <span class="key">if</span> <span class="key">not</span> <span class="nam">format</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t348" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Format required (lower case string)"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t349" class="stm mis">    <span class="key">if</span> <span class="nam">format</span> <span class="op">!=</span> <span class="nam">format</span><span class="op">.</span><span class="nam">lower</span><span class="op">(</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t350" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Format string '%s' should be lower case"</span> <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t351" class="stm mis">    <span class="key">if</span> <span class="nam">alphabet</span> <span class="key">is</span> <span class="key">not</span> <span class="key">None</span> <span class="key">and</span> <span class="key">not</span> <span class="op">(</span><span class="nam">isinstance</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">,</span> <span class="nam">Alphabet</span><span class="op">)</span> <span class="key">or</span><span class="strut">&nbsp;</span></p>
<p id="t352" class="pln">                                     <span class="nam">isinstance</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">,</span> <span class="nam">AlphabetEncoder</span><span class="op">)</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t353" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Invalid alphabet, %s"</span> <span class="op">%</span> <span class="nam">repr</span><span class="op">(</span><span class="nam">alphabet</span><span class="op">)</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t354" class="stm mis">    <span class="key">if</span> <span class="nam">seq_count</span> <span class="key">is</span> <span class="key">not</span> <span class="key">None</span> <span class="key">and</span> <span class="key">not</span> <span class="nam">isinstance</span><span class="op">(</span><span class="nam">seq_count</span><span class="op">,</span> <span class="nam">int</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t355" class="stm mis">        <span class="key">raise</span> <span class="nam">TypeError</span><span class="op">(</span><span class="str">"Need integer for seq_count (sequences per alignment)"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t356" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t357" class="stm mis">    <span class="key">with</span> <span class="nam">as_handle</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="str">'rU'</span><span class="op">)</span> <span class="key">as</span> <span class="nam">fp</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t358" class="pln">        <span class="com"># Map the file format to a sequence iterator:</span><span class="strut">&nbsp;</span></p>
<p id="t359" class="stm mis">        <span class="key">if</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">_FormatToIterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t360" class="stm mis">            <span class="nam">iterator_generator</span> <span class="op">=</span> <span class="nam">_FormatToIterator</span><span class="op">[</span><span class="nam">format</span><span class="op">]</span><span class="strut">&nbsp;</span></p>
<p id="t361" class="stm mis">            <span class="key">if</span> <span class="nam">alphabet</span> <span class="key">is</span> <span class="key">None</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t362" class="stm mis">                <span class="nam">i</span> <span class="op">=</span> <span class="nam">iterator_generator</span><span class="op">(</span><span class="nam">fp</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t363" class="pln">            <span class="key">else</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t364" class="stm mis">                <span class="key">try</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t365" class="pln">                    <span class="com"># Initially assume the optional alphabet argument is supported</span><span class="strut">&nbsp;</span></p>
<p id="t366" class="stm mis">                    <span class="nam">i</span> <span class="op">=</span> <span class="nam">iterator_generator</span><span class="op">(</span><span class="nam">fp</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">=</span><span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t367" class="stm mis">                <span class="key">except</span> <span class="nam">TypeError</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t368" class="pln">                    <span class="com"># It isn't supported.</span><span class="strut">&nbsp;</span></p>
<p id="t369" class="stm mis">                    <span class="nam">i</span> <span class="op">=</span> <span class="nam">_force_alphabet</span><span class="op">(</span><span class="nam">iterator_generator</span><span class="op">(</span><span class="nam">fp</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">)</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t370" class="pln">                                        <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t371" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t372" class="stm mis">        <span class="key">elif</span> <span class="nam">format</span> <span class="key">in</span> <span class="nam">SeqIO</span><span class="op">.</span><span class="nam">_FormatToIterator</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t373" class="pln">            <span class="com"># Exploit the existing SeqIO parser to the dirty work!</span><span class="strut">&nbsp;</span></p>
<p id="t374" class="stm mis">            <span class="nam">i</span> <span class="op">=</span> <span class="nam">_SeqIO_to_alignment_iterator</span><span class="op">(</span><span class="nam">fp</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t375" class="pln">                                             <span class="nam">alphabet</span><span class="op">=</span><span class="nam">alphabet</span><span class="op">,</span><span class="strut">&nbsp;</span></p>
<p id="t376" class="pln">                                             <span class="nam">seq_count</span><span class="op">=</span><span class="nam">seq_count</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t377" class="pln">        <span class="key">else</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t378" class="stm mis">            <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"Unknown format '%s'"</span> <span class="op">%</span> <span class="nam">format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t379" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t380" class="pln">        <span class="com"># TODO: As of Python 3.3, can write "yield from i" instead. See PEP380.</span><span class="strut">&nbsp;</span></p>
<p id="t381" class="pln">        <span class="com"># For loop imposes some overhead... wait until we drop Python 2.7 to fix it.</span><span class="strut">&nbsp;</span></p>
<p id="t382" class="stm mis">        <span class="key">for</span> <span class="nam">a</span> <span class="key">in</span> <span class="nam">i</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t383" class="stm mis">            <span class="key">yield</span> <span class="nam">a</span><span class="strut">&nbsp;</span></p>
<p id="t384" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t385" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t386" class="stm run hide_run"><span class="key">def</span> <span class="nam">read</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">=</span><span class="key">None</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">=</span><span class="key">None</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t387" class="pln">    <span class="str">"""Turn an alignment file into a single MultipleSeqAlignment object.</span><span class="strut">&nbsp;</span></p>
<p id="t388" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t389" class="pln"><span class="str">    Arguments:</span><span class="strut">&nbsp;</span></p>
<p id="t390" class="pln"><span class="str">     - handle    - handle to the file, or the filename as a string</span><span class="strut">&nbsp;</span></p>
<p id="t391" class="pln"><span class="str">       (note older versions of Biopython only took a handle).</span><span class="strut">&nbsp;</span></p>
<p id="t392" class="pln"><span class="str">     - format    - string describing the file format.</span><span class="strut">&nbsp;</span></p>
<p id="t393" class="pln"><span class="str">     - alphabet  - optional Alphabet object, useful when the sequence type</span><span class="strut">&nbsp;</span></p>
<p id="t394" class="pln"><span class="str">       cannot be automatically inferred from the file itself</span><span class="strut">&nbsp;</span></p>
<p id="t395" class="pln"><span class="str">       (e.g. fasta, phylip, clustal)</span><span class="strut">&nbsp;</span></p>
<p id="t396" class="pln"><span class="str">     - seq_count - Optional integer, number of sequences expected in each</span><span class="strut">&nbsp;</span></p>
<p id="t397" class="pln"><span class="str">       alignment.  Recommended for fasta format files.</span><span class="strut">&nbsp;</span></p>
<p id="t398" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t399" class="pln"><span class="str">    If the handle contains no alignments, or more than one alignment,</span><span class="strut">&nbsp;</span></p>
<p id="t400" class="pln"><span class="str">    an exception is raised.  For example, using a PFAM/Stockholm file</span><span class="strut">&nbsp;</span></p>
<p id="t401" class="pln"><span class="str">    containing one alignment:</span><span class="strut">&nbsp;</span></p>
<p id="t402" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t403" class="pln"><span class="str">    >>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t404" class="pln"><span class="str">    >>> filename = "Clustalw/protein.aln"</span><span class="strut">&nbsp;</span></p>
<p id="t405" class="pln"><span class="str">    >>> format = "clustal"</span><span class="strut">&nbsp;</span></p>
<p id="t406" class="pln"><span class="str">    >>> alignment = AlignIO.read(filename, format)</span><span class="strut">&nbsp;</span></p>
<p id="t407" class="pln"><span class="str">    >>> print("Alignment of length %i" % alignment.get_alignment_length())</span><span class="strut">&nbsp;</span></p>
<p id="t408" class="pln"><span class="str">    Alignment of length 411</span><span class="strut">&nbsp;</span></p>
<p id="t409" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t410" class="pln"><span class="str">    If however you want the first alignment from a file containing</span><span class="strut">&nbsp;</span></p>
<p id="t411" class="pln"><span class="str">    multiple alignments this function would raise an exception.</span><span class="strut">&nbsp;</span></p>
<p id="t412" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t413" class="pln"><span class="str">    >>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t414" class="pln"><span class="str">    >>> filename = "Emboss/needle.txt"</span><span class="strut">&nbsp;</span></p>
<p id="t415" class="pln"><span class="str">    >>> format = "emboss"</span><span class="strut">&nbsp;</span></p>
<p id="t416" class="pln"><span class="str">    >>> alignment = AlignIO.read(filename, format)</span><span class="strut">&nbsp;</span></p>
<p id="t417" class="pln"><span class="str">    Traceback (most recent call last):</span><span class="strut">&nbsp;</span></p>
<p id="t418" class="pln"><span class="str">        ...</span><span class="strut">&nbsp;</span></p>
<p id="t419" class="pln"><span class="str">    ValueError: More than one record found in handle</span><span class="strut">&nbsp;</span></p>
<p id="t420" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t421" class="pln"><span class="str">    Instead use:</span><span class="strut">&nbsp;</span></p>
<p id="t422" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t423" class="pln"><span class="str">    >>> from Bio import AlignIO</span><span class="strut">&nbsp;</span></p>
<p id="t424" class="pln"><span class="str">    >>> filename = "Emboss/needle.txt"</span><span class="strut">&nbsp;</span></p>
<p id="t425" class="pln"><span class="str">    >>> format = "emboss"</span><span class="strut">&nbsp;</span></p>
<p id="t426" class="pln"><span class="str">    >>> alignment = next(AlignIO.parse(filename, format))</span><span class="strut">&nbsp;</span></p>
<p id="t427" class="pln"><span class="str">    >>> print("First alignment has length %i" % alignment.get_alignment_length())</span><span class="strut">&nbsp;</span></p>
<p id="t428" class="pln"><span class="str">    First alignment has length 124</span><span class="strut">&nbsp;</span></p>
<p id="t429" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t430" class="pln"><span class="str">    You must use the Bio.AlignIO.parse() function if you want to read multiple</span><span class="strut">&nbsp;</span></p>
<p id="t431" class="pln"><span class="str">    records from the handle.</span><span class="strut">&nbsp;</span></p>
<p id="t432" class="pln"><span class="str">    """</span><span class="strut">&nbsp;</span></p>
<p id="t433" class="stm mis">    <span class="nam">iterator</span> <span class="op">=</span> <span class="nam">parse</span><span class="op">(</span><span class="nam">handle</span><span class="op">,</span> <span class="nam">format</span><span class="op">,</span> <span class="nam">seq_count</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t434" class="stm mis">    <span class="key">try</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t435" class="stm mis">        <span class="nam">first</span> <span class="op">=</span> <span class="nam">next</span><span class="op">(</span><span class="nam">iterator</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t436" class="stm mis">    <span class="key">except</span> <span class="nam">StopIteration</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t437" class="stm mis">        <span class="nam">first</span> <span class="op">=</span> <span class="key">None</span><span class="strut">&nbsp;</span></p>
<p id="t438" class="stm mis">    <span class="key">if</span> <span class="nam">first</span> <span class="key">is</span> <span class="key">None</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t439" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"No records found in handle"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t440" class="stm mis">    <span class="key">try</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t441" class="stm mis">        <span class="nam">second</span> <span class="op">=</span> <span class="nam">next</span><span class="op">(</span><span class="nam">iterator</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t442" class="stm mis">    <span class="key">except</span> <span class="nam">StopIteration</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t443" class="stm mis">        <span class="nam">second</span> <span class="op">=</span> <span class="key">None</span><span class="strut">&nbsp;</span></p>
<p id="t444" class="stm mis">    <span class="key">if</span> <span class="nam">second</span> <span class="key">is</span> <span class="key">not</span> <span class="key">None</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t445" class="stm mis">        <span class="key">raise</span> <span class="nam">ValueError</span><span class="op">(</span><span class="str">"More than one record found in handle"</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t446" class="stm mis">    <span class="key">if</span> <span class="nam">seq_count</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t447" class="stm mis">        <span class="key">if</span> <span class="nam">len</span><span class="op">(</span><span class="nam">first</span><span class="op">)</span> <span class="op">!=</span> <span class="nam">seq_count</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t448" class="stm mis">            <span class="key">raise</span> <span class="nam">RuntimeError</span><span class="op">(</span><span class="str">"More sequences found in alignment than specified in seq_count: %s."</span> <span class="op">%</span> <span class="nam">seq_count</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t449" class="stm mis">    <span class="key">return</span> <span class="nam">first</span><span class="strut">&nbsp;</span></p>
<p id="t450" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t451" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t452" class="stm run hide_run"><span class="key">def</span> <span class="nam">convert</span><span class="op">(</span><span class="nam">in_file</span><span class="op">,</span> <span class="nam">in_format</span><span class="op">,</span> <span class="nam">out_file</span><span class="op">,</span> <span class="nam">out_format</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">=</span><span class="key">None</span><span class="op">)</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t453" class="pln">    <span class="str">"""Convert between two alignment files, returns number of alignments.</span><span class="strut">&nbsp;</span></p>
<p id="t454" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t455" class="pln"><span class="str">    Arguments:</span><span class="strut">&nbsp;</span></p>
<p id="t456" class="pln"><span class="str">     - in_file - an input handle or filename</span><span class="strut">&nbsp;</span></p>
<p id="t457" class="pln"><span class="str">     - in_format - input file format, lower case string</span><span class="strut">&nbsp;</span></p>
<p id="t458" class="pln"><span class="str">     - output - an output handle or filename</span><span class="strut">&nbsp;</span></p>
<p id="t459" class="pln"><span class="str">     - out_file - output file format, lower case string</span><span class="strut">&nbsp;</span></p>
<p id="t460" class="pln"><span class="str">     - alphabet - optional alphabet to assume</span><span class="strut">&nbsp;</span></p>
<p id="t461" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t462" class="pln"><span class="str">    **NOTE** - If you provide an output filename, it will be opened which will</span><span class="strut">&nbsp;</span></p>
<p id="t463" class="pln"><span class="str">    overwrite any existing file without warning. This may happen if even the</span><span class="strut">&nbsp;</span></p>
<p id="t464" class="pln"><span class="str">    conversion is aborted (e.g. an invalid out_format name is given).</span><span class="strut">&nbsp;</span></p>
<p id="t465" class="pln"><span class="str">    """</span><span class="strut">&nbsp;</span></p>
<p id="t466" class="pln">    <span class="com"># TODO - Add optimised versions of important conversions</span><span class="strut">&nbsp;</span></p>
<p id="t467" class="pln">    <span class="com"># For now just off load the work to SeqIO parse/write</span><span class="strut">&nbsp;</span></p>
<p id="t468" class="stm mis">    <span class="key">with</span> <span class="nam">as_handle</span><span class="op">(</span><span class="nam">in_file</span><span class="op">,</span> <span class="str">'rU'</span><span class="op">)</span> <span class="key">as</span> <span class="nam">in_handle</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t469" class="pln">        <span class="com"># Don't open the output file until we've checked the input is OK:</span><span class="strut">&nbsp;</span></p>
<p id="t470" class="stm mis">        <span class="nam">alignments</span> <span class="op">=</span> <span class="nam">parse</span><span class="op">(</span><span class="nam">in_handle</span><span class="op">,</span> <span class="nam">in_format</span><span class="op">,</span> <span class="key">None</span><span class="op">,</span> <span class="nam">alphabet</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t471" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t472" class="pln">        <span class="com"># This will check the arguments and issue error messages,</span><span class="strut">&nbsp;</span></p>
<p id="t473" class="pln">        <span class="com"># after we have opened the file which is a shame.</span><span class="strut">&nbsp;</span></p>
<p id="t474" class="stm mis">        <span class="key">with</span> <span class="nam">as_handle</span><span class="op">(</span><span class="nam">out_file</span><span class="op">,</span> <span class="str">'w'</span><span class="op">)</span> <span class="key">as</span> <span class="nam">out_handle</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t475" class="stm mis">            <span class="nam">count</span> <span class="op">=</span> <span class="nam">write</span><span class="op">(</span><span class="nam">alignments</span><span class="op">,</span> <span class="nam">out_handle</span><span class="op">,</span> <span class="nam">out_format</span><span class="op">)</span><span class="strut">&nbsp;</span></p>
<p id="t476" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t477" class="stm mis">    <span class="key">return</span> <span class="nam">count</span><span class="strut">&nbsp;</span></p>
<p id="t478" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t479" class="pln"><span class="strut">&nbsp;</span></p>
<p id="t480" class="stm run hide_run"><span class="key">if</span> <span class="nam">__name__</span> <span class="op">==</span> <span class="str">"__main__"</span><span class="op">:</span><span class="strut">&nbsp;</span></p>
<p id="t481" class="stm mis">    <span class="key">from</span> <span class="nam">Bio</span><span class="op">.</span><span class="nam">_utils</span> <span class="key">import</span> <span class="nam">run_doctest</span><span class="strut">&nbsp;</span></p>
<p id="t482" class="stm mis">    <span class="nam">run_doctest</span><span class="op">(</span><span class="op">)</span><span class="strut">&nbsp;</span></p>

            </td>
        </tr>
    </table>
</div>

<div id="footer">
    <div class="content">
        <p>
            <a class="nav" href="index.html">&#xab; index</a> &nbsp; &nbsp; <a class="nav" href="https://coverage.readthedocs.io">coverage.py v4.5.2</a>,
            created at 2018-11-30 03:46
        </p>
    </div>
</div>

</body>
</html>
